master/internal/api_experiment.go

```diff
@@ -46,6 +46,60 @@ import (
 
 var experimentsAddr = actor.Addr("experiments")
 
+// Catches information on active running trials.
+type experimentAllocation struct {
+	JobID       *string
+	NumRunning  int32
+	NumStarting int32
+	TaskID      *string
+}
+
+func (a *apiServer) enrichExperimentState(experiments []*experimentv1.Experiment) (
+	[]*experimentv1.Experiment, error,
+) {
+	// filter allocations by JobIDs on this page of experiments
+	jobFilter := []string{}
+	for _, exp := range experiments {
+		jobFilter = append(jobFilter, exp.JobId)
+	}
+
+	// get active or pending tasks
+	tasks := []*experimentAllocation{}
+	err := a.m.db.Query(
+		"get_active_allocations_by_job",
+		&tasks,
+		strings.Join(jobFilter, ","),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Collect state information by JobID
+	applyIDs := make(map[string]experimentv1.State)
+	for _, task := range tasks {
+		if task.NumRunning > 0 {
+			applyIDs[*task.JobID] = experimentv1.State_STATE_ACTIVE
+		} else if task.NumStarting > 0 {
+			_, expStoredState := applyIDs[*task.JobID]
+			if !expStoredState {
+				applyIDs[*task.JobID] = experimentv1.State_STATE_PENDING
+			}
+		}
```

Comment:
let's make this exhaustive since it is easier to reason about (in terms of my query comments). same for trial state.
```
switch {
case task.IsRunning:
	byJobID[task.JobID] = model.MostProgressedAllocationState(byJobID[task.JobID], experimentv1.State_STATE_ACTIVE)
case task.IsStarting:
	byJobID[task.JobID] = model.MostProgressedAllocationState(byJobID[task.JobID], experimentv1.State_STATE_PENDING)
default:
	byJobID[task.JobID] = model.MostProgressedAllocationState(byJobID[task.JobID], experimentv1.State_STATE_QUEUED)
}
```

Reply:
:+1: